src/main/java/com/contentgrid/gateway/security/bearer/DynamicJwtAuthenticationManagerResolver.java

```diff
@@ -0,0 +1,45 @@
+package com.contentgrid.gateway.security.bearer;
+
+import com.contentgrid.gateway.runtime.RuntimeRequestResolver;
+import com.contentgrid.gateway.security.oidc.ReactiveClientRegistrationIdResolver;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.authentication.ReactiveAuthenticationManager;
+import org.springframework.security.authentication.ReactiveAuthenticationManagerResolver;
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
+import org.springframework.security.oauth2.jwt.ReactiveJwtDecoders;
+import org.springframework.security.oauth2.server.resource.authentication.JwtReactiveAuthenticationManager;
+import org.springframework.web.server.ServerWebExchange;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+@RequiredArgsConstructor
+public class DynamicJwtAuthenticationManagerResolver implements
+        ReactiveAuthenticationManagerResolver<ServerWebExchange> {
+
+    private final RuntimeRequestResolver runtimeRequestResolver;
+    private final ReactiveClientRegistrationIdResolver reactiveClientRegistrationIdResolver;
+    private final ReactiveClientRegistrationRepository clientRegistrationRepository;
+    private final Map<String, Mono<ReactiveAuthenticationManager>> authenticationManagers = new ConcurrentHashMap<>();
+
+    @Override
+    public Mono<ReactiveAuthenticationManager> resolve(ServerWebExchange exchange) {
+
+        return Mono.justOrEmpty(this.runtimeRequestResolver.resolveApplicationId(exchange))
+                .flatMap(this.reactiveClientRegistrationIdResolver::resolveRegistrationId)
+                .flatMap(this.clientRegistrationRepository::findByRegistrationId)
+                .map(clientRegistration -> clientRegistration.getProviderDetails().getIssuerUri())
+                .flatMap(issuer -> this.authenticationManagers.computeIfAbsent(issuer,
+                        DynamicJwtAuthenticationManagerResolver::createJwtAuthenticationManager));
+    }
+
+    // Note: this is copied from Spring Security TrustedIssuerJwtAuthenticationManagerResolver
+    private static Mono<ReactiveAuthenticationManager> createJwtAuthenticationManager(String issuer) {
+        return Mono.<ReactiveAuthenticationManager>fromCallable(
+                        () -> new JwtReactiveAuthenticationManager(ReactiveJwtDecoders.fromIssuerLocation(issuer)))
+                .subscribeOn(Schedulers.boundedElastic())
```

Comment:
This looks extremely weird that it's forced on a certain scheduler, and then some kind of forever-cache is put after it?

Given that the Mono<AuthenticationManager> is already kept in a field, wouldn't it already be cached there?

Reply:
1. Project reactor recommends wrapping legacy blocking IO code in a `Mono.fromCallable(...)` and scheduling this on `Schedulers.boundedElastic()` - see https://projectreactor.io/docs/core/release/reference/#faq.wrap-blocking
2. The cache is because it is an expensive operation: resolving the authentication manager means calling out to the OIDC metadata discovery endpoint. We definately don't want to do this for every incoming request to the gateway
3. the cache TTL are defaults that are copied from Spring Security `JwtIssuerReactiveAuthenticationManagerResolver` implementation - values used from the metadata discovery are the Keycloak realm authorization endpoint and token endpoint. I'm not expecting those to change over the lifetime of the realm at this moment, there is no advantage in expiring those cache entries right now
4. I don't think you can re-use (a callable-backed) a non-cached `Mono<>`